File: src/main/java/codereview/simpleorder/domain/item/Item.java
package codereview.simpleorder.domain.item;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
@ToString
public class Item {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "item_id")
    private Long id;

    @Column(nullable = false, length = 100, unique = true)
    private String name;

    @Column(nullable = false)
    private String size;

    @Column(nullable = false)
    private int price;

    @Column(nullable = false)
    private int availableQuantity;

    @Version
    private Long version;

    public Item(String name, String size, int price, int availableQuantity) {
        this.name = name;
        this.size = size;
        this.price = price;
        this.availableQuantity = availableQuantity;
    }

    public void decreaseQuantity(int orderQuantity) {
        validateCanDecrease(orderQuantity);
        this.availableQuantity -= orderQuantity;
    }

    private void validateCanDecrease(int orderQuantity) {
        if (availableQuantity - orderQuantity < 0) {
            throw new IllegalArgumentException("주문수량에 비해 상품의 재고수량이 충분하지 않습니다.");
        }
    }
}

File: item/src/main/kotlin/msa/with/ddd/item/domain/Item.kt
package msa.with.ddd.item.domain

import jakarta.persistence.*
import lombok.AccessLevel
import lombok.Getter
import lombok.NoArgsConstructor
import lombok.ToString

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
@ToString
class Item(
    @field:Column(
        nullable = false,
        length = 100,
        unique = true
    ) val name: String,

    @field:Column(nullable = false) val size: String,

    @field:Column(nullable = false) val price: Int,

    @field:Column(nullable = false) var availableQuantity: Int
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "item_id")
    val id: Long? = null

    @Version
    val version: Long? = null

    // default constructor for using JPA
    constructor() : this("", "", 0, 0)

    // Size가 존재하지 않는 경우
    constructor(name: String, price: Int, availableQuantity: Int) : this(name, "-", price, availableQuantity)

    fun decreaseQuantity(orderQuantity: Int) {
        validateCanDecrease(orderQuantity)
        availableQuantity -= orderQuantity
    }

    private fun validateCanDecrease(orderQuantity: Int) {
        check(availableQuantity - orderQuantity >= 0) { "주문수량에 비해 상품의 재고수량이 충분하지 않습니다." }
    }
}